plugins/grafana-custom-plugins/grafana-http-table-plugin/src/components/TablePanel.test.tsx

```diff
@@ -0,0 +1,46 @@
+import { configure, shallow } from 'enzyme';
+import Adapter from 'enzyme-adapter-react-16';
+import { TablePanel } from './TablePanel';
+import { LoadingState, PanelProps, TimeRange, toDataFrame } from '@grafana/data';
+import React from 'react';
+
+configure({ adapter: new Adapter() });
+
+describe('Table Plugin Test', () => {
+  it('Should render Table', () => {
+    let props = {} as PanelProps;
+    let timeRange = {} as TimeRange;
+    props.data = {
+      series: [
+        toDataFrame({
+          refId: 'A',
+          fields: [
+            { name: 'sourceIP', values: ['10.10.1.4'] },
+            { name: 'sourceTransportPort', values: ['42162'] },
+            { name: 'destinationIP', values: ['93.184.216.34'] },
+            { name: 'destinationTransportPort', values: ['80'] },
+            { name: 'httpVals', values: ['{"0":{"hostname":"example.com","url":"/","http_user_agent":"curl/7.88.1","http_content_type":"text/html","http_method":"GET","protocol":"HTTP/1.1","status":200,"length":1256},"1":{"hostname":"example.com","url":"/","http_user_agent":"curl/7.88.1","http_content_type":"text/html","http_method":"GET","protocol":"HTTP/1.1","status":200,"length":1256},"2":{"hostname":"example.com","url":"/","http_user_agent":"curl/7.88.1","http_content_type":"text/html","http_method":"GET","protocol":"HTTP/1.1","status":200,"length":1256}}'] },
+          ],
+        }),
+      ],
+      state: LoadingState.Done,
+      timeRange: timeRange,
+    };
+    props.width = 600,
+    props.height = 600;
+    props.options = {};
+    let component = shallow(<TablePanel {...props} />);
+    React.useLayoutEffect = React.useEffect;
+    console.log('component text: '+component.render().text());
+    let renderedHtmlString = component.render().text();
```

Comment:
I'm not familiar whether we can test it in another way, rather than converting to text. Maybe @yuntanghsu can give more constructive feedback.

But if we are just checking the text content, instead of using `includes` expression checking a series of substring ,can we just do `expect(textContent).toEqual(expectedContent);` to check the entire string?

Reply:
<img width="884" alt="image" src="https://github.com/antrea-io/theia/assets/59460118/03101fb0-c882-47f6-86cb-2d658195aede">
I think we should use `getAllByText` or `expect(getByText('10.10.1.4:42162')).toBeInTheDocument()` to check the data in the component.
For example, in your test and the picture, we can see there are 3 transaction IDs for a certain combination of source and destination. We want to make sure your component can generate exactly 3 elements, which should be based on the number of HttpVals you have. And we also want to make sure the contents are correct in these 3 elements. I'll suggest you change the vals in each transaction ID and verify the context for all of them. Otherwise the result can be a false positive case as you only use "includes" to check the data.


I also found there are lots of warning when using component.render().text():
```
at div
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at TableContainer (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/TableContainer/TableContainer.js:40:44)
        at MRT_TableContainer (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_TableContainer.tsx:13:38)
        at div
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at Paper (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/Paper/Paper.js:69:44)
        at MRT_TablePaper (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_TablePaper.tsx:11:34)
        at useRef (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_TableRoot.tsx:50:28)
        at MaterialReactTable (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/MaterialReactTable.tsx:14:3)
        at ThemeProvider (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/private-theming/node/ThemeProvider/ThemeProvider.js:39:5)
        at ThemeProvider (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/system/ThemeProvider/ThemeProvider.js:50:5)
        at ThemeProvider (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/styles/ThemeProvider.js:21:14)
        at div

      33 |     React.useLayoutEffect = React.useEffect;
      34 |     // console.log('component text: '+component.render().text());
    > 35 |     let renderedHtmlString = component.render().text();
         |                                        ^
      36 |     expect(renderedHtmlString.includes('10.10.1.4:42162')).toEqual(true);
      37 |     expect(renderedHtmlString.includes('93.184.216.34:80')).toEqual(true);
      38 |     expect(renderedHtmlString.includes('example.com')).toEqual(true);

      at printWarning (node_modules/react-dom/cjs/react-dom-server.node.development.js:82:30)
      at error (node_modules/react-dom/cjs/react-dom-server.node.development.js:58:5)
      at Object.useLayoutEffect (node_modules/react-dom/cjs/react-dom-server.node.development.js:1617:5)
      at useLayoutEffect (node_modules/react/cjs/react.development.js:1524:21)
      at useEventCallback (node_modules/@mui/utils/useEventCallback/useEventCallback.js:20:34)
      at useRippleHandler (node_modules/@mui/material/node/ButtonBase/ButtonBase.js:160:42)
      at Object.ButtonBase [as render] (node_modules/@mui/material/node/ButtonBase/ButtonBase.js:186:22)
      at ReactDOMServerRenderer.render (node_modules/react-dom/cjs/react-dom-server.node.development.js:3872:44)
      at ReactDOMServerRenderer.read (node_modules/react-dom/cjs/react-dom-server.node.development.js:3690:29)
      at Object.renderToStaticMarkup (node_modules/react-dom/cjs/react-dom-server.node.development.js:4314:27)
      at Object.renderToStaticMarkup [as render] (node_modules/enzyme-adapter-react-16/src/ReactSixteenAdapter.js:835:31)
      at ShallowWrapper.render (node_modules/enzyme/src/ShallowWrapper.js:1108:23)
      at ShallowWrapper.call [as single] (node_modules/enzyme/src/ShallowWrapper.js:1654:21)
      at ShallowWrapper.single [as html] (node_modules/enzyme/src/ShallowWrapper.js:1104:17)
      at ShallowWrapper.html [as render] (node_modules/enzyme/src/ShallowWrapper.js:1120:23)
      at Object.render (src/components/TablePanel.test.tsx:35:40)

  console.error
    Warning: useLayoutEffect does nothing on the server, because its effect cannot be encoded into the server renderer's output format. This will lead to a mismatch between the initial, non-hydrated UI and the intended UI. To avoid this, useLayoutEffect should only be used in components that render exclusively on the client. See https://reactjs.org/link/uselayouteffect-ssr for common fixes.
        at ButtonBase (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/ButtonBase/ButtonBase.js:95:44)
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at TableSortLabel (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/TableSortLabel/TableSortLabel.js:102:44)
        at span
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at Badge (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/Badge/Badge.js:169:44)
        at Tooltip (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/Tooltip/Tooltip.js:226:44)
        at MRT_TableHeadCellSortLabel (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/head/MRT_TableHeadCellSortLabel.tsx:14:3)
        at div
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at Box (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/system/createBox.js:31:41)
        at div
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at Box (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/system/createBox.js:31:41)
        at th
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at TableCell (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/TableCell/TableCell.js:108:44)
        at MRT_TableHeadCell (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/head/MRT_TableHeadCell.tsx:20:37)
        at tr
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at TableRow (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/TableRow/TableRow.js:70:44)
        at MRT_TableHeadRow (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/head/MRT_TableHeadRow.tsx:20:3)
        at thead
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at TableHead (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/TableHead/TableHead.js:44:44)
        at MRT_TableHead (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/head/MRT_TableHead.tsx:14:3)
        at table
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at Table (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/Table/Table.js:61:44)
        at MRT_Table (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_Table.tsx:19:29)
        at div
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at TableContainer (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/TableContainer/TableContainer.js:40:44)
        at MRT_TableContainer (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_TableContainer.tsx:13:38)
        at div
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at Paper (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/Paper/Paper.js:69:44)
        at MRT_TablePaper (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_TablePaper.tsx:11:34)
        at useRef (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_TableRoot.tsx:50:28)
        at MaterialReactTable (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/MaterialReactTable.tsx:14:3)
        at ThemeProvider (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/private-theming/node/ThemeProvider/ThemeProvider.js:39:5)
        at ThemeProvider (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/system/ThemeProvider/ThemeProvider.js:50:5)
        at ThemeProvider (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/styles/ThemeProvider.js:21:14)
        at div
```

@Dhruv-J Do you have any idea what happened?

Reply:
Yeah, the material-ui library [recommends](https://mui.com/material-ui/guides/testing/) running unit testing without the use of material-ui components in order to keep code immune to breaking due to changes within material-ui internal structure. So I can't directly check for including the table itself, so I check for the string being rendered within the html itself. @yuntanghsu I can attempt using those again, but I remember when I tried using similar solutions that false would be returned even though the text was supposedly being rendered. The warnings happen because the render() function is being called in a unit test, and so the result isn't actually visible anywhere. I looked into this issue specifically with material-ui and found that it is quite common, so people largely suppress the warning or haven't found a workaround. I'll try the clause that you mentioned and update the results within this thread.